effort_controllers/src/joint_group_position_controller.cpp

```diff
@@ -0,0 +1,235 @@
+      if (!joint_urdf)
+      {
+        ROS_ERROR("Could not find joint '%s' in urdf", joint_names_[i].c_str());
+        return false;
+      }
+      joint_urdfs_.push_back(joint_urdf);
+
+      // Load PID Controller using gains set on parameter server
+      if (!pid_controllers_[i].init(ros::NodeHandle(n, joint_names_[i] + "/pid")))
+      {
+        ROS_ERROR_STREAM("Failed to load PID parameters from " << joint_names_[i] + "/pid");
+        return false;
+      }
+    }
+
+    commands_buffer_.writeFromNonRT(std::vector<double>(n_joints_, 0.0));
+
+    sub_command_ = n.subscribe<std_msgs::Float64MultiArray>("command", 1, &JointGroupPositionController::commandCB, this);
+    return true;
+  }
+
+  void JointGroupPositionController::update(const ros::Time& time, const ros::Duration& period)
+  {
+    std::vector<double> & commands = *commands_buffer_.readFromRT();
+    for(unsigned int i=0; i<n_joints_; i++)
+    {
+        double command_position = commands[i];
+//        double command_velocity = commands[i].velocity_;
+//        bool has_velocity_ =  commands[i].has_velocity_;
+
+        double error; //, vel_error;
+        double commanded_effort;
+
+        double current_position = joints_[i].getPosition();
+
+//TODO not sure if needed
+        // Make sure joint is within limits if applicable
+        //enforceJointLimits(command_position);
+
+        // Compute position error
+        if (joint_urdfs_[i]->type == urdf::Joint::REVOLUTE)
+        {
+         angles::shortest_angular_distance_with_limits(
+            current_position,
+            command_position,
+            joint_urdfs_[i]->limits->lower,
+            joint_urdfs_[i]->limits->upper,
+            error);
+        }
+        else if (joint_urdfs_[i]->type == urdf::Joint::CONTINUOUS)
+        {
+          error = angles::shortest_angular_distance(current_position, command_position);
+        }
+        else //prismatic
+        {
+          error = command_position - current_position;
+        }
+/*
+        // Decide which of the two PID computeCommand() methods to call
+        if (has_velocity_)
+        {
+          // Compute velocity error if a non-zero velocity command was given
+          vel_error = command_velocity - joint_.getVelocity();
+
+          // Set the PID error and compute the PID command with nonuniform
+          // time step size. This also allows the user to pass in a precomputed derivative error.
+          commanded_effort = pid_controller_.computeCommand(error, vel_error, period);
+        }
+        else
+        {
+*/
+          // Set the PID error and compute the PID command with nonuniform
+          // time step size.
+          commanded_effort = pid_controllers_[i].computeCommand(error, period);
+//        }
+
+        joints_[i].setCommand(commanded_effort);
+    }
+
+    /*
+    // publish state
+    if (loop_count_ % 10 == 0)
+    {
+      if(controller_state_publisher_ && controller_state_publisher_->trylock())
```

Comment:
Do we want to have a controller state published?
Then we would need a new message type for it.